### ISSUE TITLE
ParameterName fails in certain cases where two classes reside in one file

### ISSUE THREAD

**Issue by nicktrav** [google/error-prone#792]
In the case of multiple classes residing in the same .java file, say `Foo` (the public class) and `Bar` in the file `Foo.java`, if a second class `Baz`, residing in file `Baz.java` extends `Foo`, error-prone can fail in what appears to be the case when `Baz.java` is compiled and checked before `Foo.java`.

Foo.java:

```java
class Bar {
}

public class Foo {

  public void acceptInteger(Integer integer) {
  }

  public void callAccept() {
    acceptInteger(0);
  }
}
```

Baz.java:

```java
public class Baz extends Foo {
}
```

Specifically, the issue is in the `ParameterName` check, which is failing with an `IndexOutOfBoundsException` [here](https://github.com/google/error-prone/blob/v2.1.2/core/src/main/java/com/google/errorprone/bugpatterns/ParameterName.java#L85).

[Here's a link](https://github.com/nicktrav/error-prone/commit/a2be36b2bf563edd64d326cd4c57d8f62a190944) to a commit reproducing the issue.

Running the following test case:

```
mvn test -pl core -Dtest='com.google.errorprone.bugpatterns.MultipleClassWithExtendsTest'
```

leads to the following failure for one of the two tests:

```
Failed tests:
  MultipleClassWithExtendsTest.failingTest:40 An unhandled exception was thrown by the Error Prone static analysis plugin.
     Please report this at https://github.com/google/error-prone/issues/new and include the following:

     error-prone version: unknown version
     Stack Trace:
     java.lang.IndexOutOfBoundsException
        at java.nio.HeapCharBuffer.subSequence(HeapCharBuffer.java:580)
        at java.nio.HeapCharBuffer.subSequence(HeapCharBuffer.java:42)
        at com.google.errorprone.bugpatterns.ParameterName.checkArguments(ParameterName.java:85)
        at com.google.errorprone.bugpatterns.ParameterName.matchMethodInvocation(ParameterName.java:64)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:907)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:146)
        at com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1644)
     ... snip ...
```

Affects v2.1.2, but also confirmed to exist at HEAD on master.

**Comment by cushon** [comment]
I suspect this is #781. Does adding the javac flag `-XDcompilePolicy=simple` work around the crash?

**Comment by nicktrav** [comment]
Ah damn. Sorry, I should have searched harder! This was a gnarly one to re-pro.

What's interesting is that the tests in the commit I provided still fail on master. I'm running with `mvn clean test`. I wonder if that's something that needs to be tweaked in the maven settings, or if it's something different (some other edge-case that we're tickling perhaps?).

I see there was some change in https://github.com/google/error-prone/commit/73e24807d999832e694367242306ec07d62bb5a6 to the flags.

Feel free to include the test cases. Or I can open a PR.

**Comment by cushon** [comment]
73e24807d999832e694367242306ec07d62bb5a6 hadn't been merged into master yet. Your test was still failing after I merged it when I ran it by itself with `mvn surefire:test -pl core -Dtest=MultipleClassWithExtendsTest`, but running all of the tests with `mvn test` worked. Running the test by itself worked after a `mvn install` - based on the line numbers in the stack trace I think it was seeing the wrong version of `test_helpers`.

Anyway, I'm confident this is #781. Thanks for finding that repro, I had be happy to have more test coverage for this if you want to turn it into a PR.
